src/rulenode.jl

```diff
@@ -1,27 +1,41 @@
+"""
+Type for representing expression trees.
+"""
+abstract type AbstractRuleNode end
+
 
 """
 RuleNode
 Type for representing nodes in an expression tree.
 """
-mutable struct RuleNode
+mutable struct RuleNode <: AbstractRuleNode
 	ind::Int # index in grammar
 	_val::Any  #value of _() evals
-	children::Vector{RuleNode}
+	children::Vector{AbstractRuleNode}
+end
+
+mutable struct Hole <: AbstractRuleNode
+	domain::BitVector
 end
 
 RuleNode(ind::Int) = RuleNode(ind, nothing, RuleNode[])
+RuleNode(ind::Int, children::Vector{AbstractRuleNode}) = RuleNode(ind, nothing, children)
 RuleNode(ind::Int, children::Vector{RuleNode}) = RuleNode(ind, nothing, children)
+RuleNode(ind::Int, children::Vector{Hole}) = RuleNode(ind, nothing, children)
 RuleNode(ind::Int, _val::Any) = RuleNode(ind, _val, RuleNode[])
```

Comment:
should `RuleNode[]` be turned into a list of holes?

Same for line 12?

Reply:
I agree, I turned it into AbstractRuleNode lists.
To actually fill the child list with holes, we need access to the grammar to get the number of children and their types.
Therefore, I created separate constructors for this and kept these ones intact.